### DISCUSSION TITLE
Query batching + Row Level Security

### DISCUSSION THREAD

**Discussion by A-Shleifman** [prisma/prisma#11824] [Q&A]
I have Row Level Security configured at the DB level. The DB decides which data should be available to the current user based on transaction variables I'm configuring in Prisma middleware:

```ts
export const rlsMiddleware: Prisma.Middleware = async (params, next) => {
  if (params.runInTransaction || !ctx().user || !ctx().tenantId || typeof params.model !== 'string') {
    return next(params);
  }

  const modelName = params.model.charAt(0).toLowerCase() + params.model.slice(1);

  const statements = [
    prisma.$executeRawUnsafe(`SET LOCAL rls.tenantId = '${ctx().tenantId}';`),
    prisma.$executeRawUnsafe(`SET LOCAL rls.userId = '${ctx().user.id}';`),
    // @ts-ignore
    prisma[modelName][params.action](params.args),
  ];

  const path = params.dataPath.at(-1);

  const results = (await prisma.$transaction(statements)) as unknown as object[];
  const result = results.at(-1)!;

  return path ? result[path as keyof typeof result] : result;
};
```

This does work very well. The only problem is that after converting queries to transactions (to set the variables) automatic Prisma batching is no longer working, so GraphQL resolvers are making N+1 requests. I could resolve this problem by writing custom dataloaders, but ideally, I would like to avoid that. Is there a way to run a middleware after batching? Any other way to achieve batching while still using RLS?

**Comment by nurul3101** [comment]
Hey @A-Shleifman 👋 
You should checkout [Cerbos](https://cerbos.dev/) they abstract authorisation while still allowing granular access control to achieve RLS. 
They do have integration with Prisma as mentioned in their docs [here](https://cerbos.dev/ecosystem/cerbos-prisma).
Delegating RLS to cerbos might help you in this case.

**Reply by A-Shleifman** [reply]
This looks very interesting! Thank you for the hint @nurul3101!

**Reply by eduhenke** [reply]
I don't understand how cerbos can be used to plan a query? For example if i'm listing all the users in my tenant, how would cerbos help with that?

**Reply by eduhenke** [reply]
also @A-Shleifman, could you figure out a way to do RLS + Query batching? I'm also trying to figure it out

**Reply by A-Shleifman** [reply]
I still didn't have time to try Cerbos, but I guess the idea is that Cerbos takes over authorisation, which allows using prisma without the middleware above, so batching should work properly.

**Reply by eduhenke** [reply]
Yeah, from what I saw on the example, it only works with prisma as a allow/deny server. We manually pass the entire resource and access metadata(which user is accessing, its role, its organization, etc.) to Cerbos and it either allows or denies access to that resource, which doesn't at all replace RLS. Like `const product = await prisma.product.find(); const bool = cerbos.allow(product, user);`

I've seen that this Cerbos can return a query plan, but how to translate this query plan on a specific format to a Prisma query, is undocumented and would be really error-prone for us(simple users) to write.

---

Nevertheless, @A-Shleifman were you able to solve the problem of RLS + query batching in some way?

**Reply by A-Shleifman** [reply]
Just watched Cerbos's video and yes, maybe it's not a solution. 

I had to write a custom dataloader for each findUnique to batch them in a single findMany request. It works well, but requires extra code and missing a dataloader leads to N requests being made. 

I had to replace requests like `prisma.users.findUnique(id).settings` with 2 separate requests to batch them.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
